internal/sink/sink.go

```diff
@@ -225,43 +229,65 @@ func (whs WebhookSink) Send(secret, acceptType string, msg *wrp.Message) error {
 func (whs WebhookSink) batchMsgs() {
 	msgs := []*wrp.Message{}
 	ticker := time.NewTicker(whs.BatchLinger) //TODO: setting to this for now - open to suggestions
-	select {
-	case msg := <-whs.ch:
-		msgs = append(msgs, msg)
-		if len(msgs) == whs.BatchMessages {
-			whs.sendBatch(msgs)
+	expired := false
+	var ready <-chan struct{}
+	for {
+		select {
+		case <-ready:
+			ready = nil
+		case msg, ok := <-whs.ch:
+			if !ok {
+				return
+			}
+			msgs = append(msgs, msg)
+		case <-ticker.C:
+			expired = true
+			ticker = time.NewTicker(whs.BatchLinger)
+		}
+		if (expired || len(msgs) == whs.BatchMessages) && ready == nil{
+			ready = whs.sendBatch(msgs)
 		}
-	case <-ticker.C:
-		whs.sendBatch(msgs)
-		//TODO: do we need to set a new ticker?
 	}
+
 }
 
-func (whs WebhookSink) sendBatch(msgs []*wrp.Message) {
-	var errs error
-	if len(*whs.Hash) == len(whs.webooks) {
-		//TODO: flush out the error handling for kafka
-		for _, msg := range msgs {
-			if v2, ok := whs.webooks[whs.Hash.Get(GetKey(whs.HashField, msg))]; ok {
-				err := v2.send(v2.secret, v2.acceptType, msg)
-				if err != nil {
-					errs = errors.Join(errs, err)
+/*
+need to consider:
+ 1. what part of sendBatch should be done async?
+ 2. where should rate limiting happen? where is flow control happening? (workers.Acquire)
+    2a. identify potential resource starvation (i.e. running too many threads than CPU cores)
+    2b. use worker interface (aquire & release) to enforce rate limiting
+    2b1. where should worker.Acquire be called and where should worker.Release be called
+
+3. how do we define sendBatch to be ready. what conditions signal channel to be closed/is ready?
+*/
+
+func (whs WebhookSink) sendBatch(msgs []*wrp.Message) <-chan struct{} {
+	ready := make(chan struct{})
+	defer close(ready)
```

Comment:
we need to add some sort of condition that determines when to call `close(ready)`. Currently, exiting `sendBatch` triggers it and that's not really doing thing useful.